src/gui/log/logpeerview.cpp

```diff
@@ -0,0 +1,91 @@
+/*
+ * Bittorrent Client using Qt and libtorrent.
+ * Copyright (C) 2019  sledgehammer999 <hammered999@gmail.com>
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#include "logpeerview.h"
+
+#include <QApplication>
+#include <QClipboard>
+#include <QHeaderView>
+#include <QKeyEvent>
+#include <QMenu>
+
+#include "logpeermodel.h"
+#include "uithememanager.h"
+
+LogPeerView::LogPeerView(QWidget *parent)
+    : QListView(parent)
+{
+    setUniformItemSizes(true);
+
+    m_listModel = new LogPeerModel(this);
+    setModel(m_listModel);
+
+    // Visual settings
+    setSelectionMode(QAbstractItemView::ExtendedSelection);
+
+#if defined(Q_OS_MAC)
+    setAttribute(Qt::WA_MacShowFocusRect, false);
+#endif
+
+    setContextMenuPolicy(Qt::CustomContextMenu);
+    connect(this, &QWidget::customContextMenuRequested, this, &LogPeerView::displayListMenu);
+}
+
+void LogPeerView::displayListMenu(const QPoint &pos)
+{
+    QMenu listMenu(this);
+    QAction *copyAct = new QAction(UIThemeManager::instance()->getIcon("edit-copy"), tr("Copy"), &listMenu);
+    QAction *clearAct = new QAction(UIThemeManager::instance()->getIcon("edit-clear"), tr("Clear"), &listMenu);
+
+    listMenu.addAction(copyAct);
+    listMenu.addAction(clearAct);
+    const QAction * const res = listMenu.exec(mapToGlobal(pos));
```

Comment:
We shouldn't use `exec()` anymore, take a look at `void TrackerFiltersList::showMenu(const QPoint &)` to see how to do it in an async way.

Reply:
should I combine both views under `BasicLogView`, I mean all they have different is model, which can be added seperately

Reply:
As I said before you can use generic View. All things you're doing in view constructors you can apply from outside it. Seems one exception is key press event handler, so you can implement generic CopiableListView and use it with your models (although you can use basic ListView and set external event handler).

Reply:
both have the same key events (is am missing something)

I was doing it this way implement `LogListView`, set model and filter seperately on  executionlogwidget, they both seem to have only this much difference

Reply:
problem is Model, they seem to do the same think but I can't seem to find an acceptable way to combine them.

Reply:
Seems the only way is using templates.

Reply:
Although seems it isn't good idea to use templates with QObject based types so we can leave two models.

Reply:
signals will not work template

how about this 
```
class BaseLogModel : public QAbstractListModel 
{
    Q_OBJECT
    Q_DISABLE_COPY(BaseLogModel)

public:
    explicit BaseLogModel(QObject *parent = nullptr);

    int rowCount(const QModelIndex &parent = {}) const override;
    int columnCount(const QModelIndex &parent = {}) const override;
    QVariant data(const QModelIndex &index, int role = Qt::DisplayRole) const override;
    void reset();

protected:
    struct Message {
        QString text;
        QColor foregroundColor;
    };
    void addMessage(const Message& msg);

private:
    QVector<Message> m_msgs;
};
```

but it will duplicate data

Reply:
>but it will duplicate data

What do you mean?

Also vector isn't good choice for collecting messages since when it is full each new message will need its shifting.

Reply:
> What do you mean?

We currently directly query the message from the logger with this we're maintaining our own  message queue.
 

> Also vector isn't good choice for collecting messages since when it is full each new message will need its shifting.

stack?

Reply:
I think you were saying to limit the number of messages in msglist, so may be this can be considered with this we're caching the message also as you requested.

Reply:
>We currently directly query the message from the logger 

Currently in this PR, I suppose...
>with this we're maintaining our own message queue.

Yes. But already converted in display friendly form. Otherwise we need to obtain and convert it regularly.

Reply:
well now I'm implementing `BaseLogModel`, but just for the discussion, isn't it will create performance implication on a large number of messages. which was the original problem. 
also, the display functions are only called for the items in the view so it can't be a much of a problem

neither std::stack or QStack provides iteration so I'm using QVector as a stack,

Reply:
> but just for the discussion, isn't it will create performance implication on a large number of messages. which was the original problem

We need to make a choice.
Seems we need to compare performance of both ways.

Anyway both models contain many duplicated code so BaseLogModel will probably relevant in both cases.

Reply:
>neither std::stack or QStack provides iteration so I'm using QVector as a stack,

There are "circular buffer" types. As far as I remember @sledgehammer999 tried to use one of them in Logger (in original PR).

Reply:
used QCache